akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.stream.scaladsl.{ Keep, OperationAttributes }
+import akka.stream.{ InPort, OutPort, Shape, EmptyShape, AmorphousShape }
+import org.reactivestreams.{ Subscription, Publisher, Subscriber }
+
+/**
+ * INTERNAL API
+ */
+private[akka] object StreamLayout {
+
+  // TODO: Materialization order
+  // TODO: Special case linear composites
+  // TODO: Cycles
+
+  sealed trait MaterializedValueNode
+  case class Combine(f: (Any, Any) ⇒ Any, dep1: MaterializedValueNode, dep2: MaterializedValueNode) extends MaterializedValueNode
+  case class Atomic(module: Module) extends MaterializedValueNode
+  case class Transform(f: Any ⇒ Any, dep: MaterializedValueNode) extends MaterializedValueNode
+  case object Ignore extends MaterializedValueNode
+
+  case class Mapping(module: Module, inPorts: Map[InPort, InPort], outPorts: Map[OutPort, OutPort])
+
+  trait Module {
+    def shape: Shape
+
+    final lazy val inPorts: Set[InPort] = shape.inlets.toSet
+    final lazy val outPorts: Set[OutPort] = shape.outlets.toSet
+
+    def isRunnable: Boolean = inPorts.isEmpty && outPorts.isEmpty
+    def isSink: Boolean = (inPorts.size == 1) && outPorts.isEmpty
+    def isSource: Boolean = (outPorts.size == 1) && inPorts.isEmpty
+    def isFlow: Boolean = (inPorts.size == 1) && (outPorts.size == 1)
+
+    def connect[A, B](from: OutPort, to: InPort): Module = {
+      require(outPorts(from), s"The output port [$from] is not part of the underlying graph.")
+      require(inPorts(to), s"The input port [$to] is not part of the underlying graph.")
+
+      CompositeModule(
+        subModules,
+        AmorphousShape(shape.inlets.filterNot(_ == to), shape.outlets.filterNot(_ == from)),
+        downstreams.updated(from, to),
+        upstreams.updated(to, from),
+        materializedValueComputation,
+        carbonCopy = () ⇒ {
+          val mapping = this.carbonCopy()
+          mapping.copy(module = mapping.module.connect(mapping.outPorts(from), mapping.inPorts(to)))
```

Comment:
everything around `carbonCopy()` and `Mapping` will very soon be gone, gone with the wind ;-)